File: Python/problem037.py
#! /usr/bin/env python3
"""
Problem 37 - Project Euler
http://projecteuler.net/index.php?section=problems&id=037
"""
from mathtools import isprime


def is_truncatable(x):
    xr = x // 10
    d = 10
    while xr > 0:
        xl = x % d
        if not (isprime(xr) and isprime(xl)):
            return False
        xr = xr // 10
        d *= 10
    return isprime(x)


if __name__ == "__main__":
    a = []
    i = 11
    while len(a) < 11:
        if is_truncatable(i):
            a.append(i)
        i += 2
    print(a)
    print(sum(a))

File: Ruby/problem037.rb
#! /usr/bin/env ruby
=begin
Problem 37 - Project Euler
http://projecteuler.net/index.php?section=problems&id=37
=end
require 'prime'
class Integer
    def truncatable?
        return false unless self.prime?
        x = y = self
        i = 1
        while x / (10 ** i) > 0
            return false unless (x / (10 ** i)).prime? && (y % (10 ** i)).prime?
            i += 1
        end
        true
    end
end

puts (11..).lazy.select(&:truncatable?).first(11).sum
